token/program-2022/src/extension/memo_transfer/processor.rs

```diff
@@ -0,0 +1,140 @@
+        },
+        processor::Processor,
+        state::Account,
+    },
+    solana_program::{
+        account_info::{next_account_info, AccountInfo},
+        entrypoint::ProgramResult,
+        msg,
+        program::invoke,
+        pubkey::Pubkey,
+        rent::Rent,
+        system_instruction,
+        sysvar::Sysvar,
+    },
+};
+
+fn process_require_memo_transfers(program_id: &Pubkey, accounts: &[AccountInfo]) -> ProgramResult {
+    let account_info_iter = &mut accounts.iter();
+    let token_account_info = next_account_info(account_info_iter)?;
+    let payer_info = next_account_info(account_info_iter)?;
+    let system_program_info = next_account_info(account_info_iter)?;
+    let owner_info = next_account_info(account_info_iter)?;
+    let owner_info_data_len = owner_info.data_len();
+
+    let mut account_data = token_account_info.data.borrow_mut();
+    let mut account = StateWithExtensionsMut::<Account>::unpack(&mut account_data)?;
+
+    Processor::validate_owner(
+        program_id,
+        &account.base.owner,
+        owner_info,
+        owner_info_data_len,
+        account_info_iter.as_slice(),
+    )?;
+
+    if let Ok(extension) = account.get_extension_mut::<MemoTransfer>() {
+        extension.require_incoming_transfer_memos = true.into();
+        Ok(())
+    } else {
+        let realloc_needed = account.realloc_needed(ExtensionType::MemoTransfer)?;
+        if let Some(diff) = realloc_needed {
+            let needs_account_type_init = account.account_type_is_empty();
+
+            // drop existing mutable borrows
+            drop(account);
+            drop(account_data);
+
+            // reallocate
+            let new_account_len = token_account_info.data_len().saturating_add(diff);
+            token_account_info.realloc(new_account_len, false)?;
+
+            // figure out if lamports needed to remain rent-exempt
+            let rent = Rent::get()?;
+            let new_minimum_balance = rent.minimum_balance(new_account_len);
+            let lamports_diff = new_minimum_balance.saturating_sub(token_account_info.lamports());
+            invoke(
+                &system_instruction::transfer(
+                    payer_info.key,
+                    token_account_info.key,
+                    lamports_diff,
+                ),
+                &[
+                    payer_info.clone(),
+                    token_account_info.clone(),
+                    system_program_info.clone(),
+                ],
+            )?;
+
+            // reborrow account state
+            let mut account_data = token_account_info.data.borrow_mut();
+            let mut account = if needs_account_type_init {
+                StateWithExtensionsMut::<Account>::unpack_after_realloc(&mut account_data)
+            } else {
+                StateWithExtensionsMut::<Account>::unpack(&mut account_data)
+            }?;
+            let extension = account.init_extension::<MemoTransfer>()?;
```

Comment:
@mvines , this isn't as clean as I originally wrote it, due to handling the mutable borrows. But should still be generalizable on T: ExtensionType

Reply:
Seems like the cleanest solution is to have `StateWithExtensionsMut` hold onto the full `AccountInfo` in the end.  Then you won't need to worry about exposing the drops and `account_type_is_empty()`

Reply:
This is super simple, now that Reallocate is its own thing